### PR TITLE
RFC: initialize `storage_proxy` early, without `messaging_service` or `gossiper`

### PR DESCRIPTION
**Pull request by kbr-**
The refactors from https://github.com/scylladb/scylla/pull/11088 allow us to easily move the initialization of `storage_proxy` early in the startup procedure, before starting `system_keyspace`, `messaging_service`, `gossiper`, `storage_service` and more.

Eventually we should be able to move initialization of `query_processor` right after `storage_proxy` (but this requires a bit of refactoring in `query_processor` too).

Only the last 7 commits are new, others come from https://github.com/scylladb/scylla/pull/11088.

### CODE REVIEW

**Comment by kbr-** [comment]
Note this commit: [main, cql_test_env: move database init earlier](https://github.com/scylladb/scylla/pull/11091/commits/42cadeaeffd866012905a8db41ecda6c8d7ae655)

I don't know why `database` was initialized so late, after `storage_service` even.
It could have already been moved earlier without my refactors.
@xemul do you know if there's a reason for initializing `database` late? Or maybe it was just that it could be moved earlier for some time now, simply nobody did this yet?

**Comment by kbr-** [comment]
@kostja regarding your comment here: https://github.com/scylladb/scylla/pull/11088#issuecomment-1190519551

> Even if such conditional query is executed against a local table (i.e. RF=1) the call chain will end up invoking an RPC, such as send_paxos_prepare(), and thus won't function without a remote() instance.

So:
- I executed dev tests locally on this branch and they pass.
- `cql_test_env` does not call `init_messaging_service` which now creates the `remote` object.
- Thus tests are passing without the `remote` object.

File: service/storage_proxy.cc [hunk 1]
```diff
@@ -2570,7 +2578,28 @@ storage_proxy::storage_proxy(distributed<replica::database>& db, gms::gossiper&
...
+    // during the entire lifetime of `storage_proxy`, which is larger than `remote`).
+    //
+    // If there's a bug though, fail the query instead of crashing Scylla.
+    //
+    // In the future we may want to introduce a 'recovery mode' in which Scylla starts
+    // without contacting the cluster and allows the user to perform local queries (say,
+    // to system tables), then this code path would be expected to happen if the user
+    // tries a remote query in this recovery mode.
+    slogger.error("attempted to access `remote` when it's uninitialized");
+    throw std::runtime_error{"Remote queries not available"};
```
Comment (psarna) [inline service/storage_proxy.cc:2602]: we have `on_internal_error` for that purpose, it logs the error too, but has an advantage of throwing in production, while aborting in test environments, making it easier to spot the problem

**Comment by xemul** [comment]
> Note this commit: [main, cql_test_env: move database init earlier](https://github.com/scylladb/scylla/pull/11091/commits/42cadeaeffd866012905a8db41ecda6c8d7ae655)
> 
> I don't know why `database` was initialized so late, after `storage_service` even. It could have already been moved earlier without my refactors. @xemul do you know if there's a reason for initializing `database` late? Or maybe it was just that it could be moved earlier for some time now, simply nobody did this yet?

@kbr- , it's just been there forever, nobody moved this yet

**Comment by xemul** [comment]
Last 7 patches -- LGTM :+1:

File: service/storage_proxy.cc [hunk 2]
```diff
@@ -3190,7 +3219,7 @@ storage_proxy::mutate_atomically(std::vector<mutation> mutations, db::consistenc
             .then(utils::result_into_future<result<>>);
 }
 
-static inet_address_vector_replica_set endpoint_filter(const gms::gossiper& g, const sstring& local_rack, const std::unordered_map<sstring, std::unordered_set<gms::inet_address>>& endpoints) {
+static inet_address_vector_replica_set endpoint_filter(const gms::gossiper* g, const sstring& local_rack, const std::unordered_map<sstring, std::unordered_set<gms::inet_address>>& endpoints) {
```
Comment (gleb-cloudius) [inline service/storage_proxy.cc:3226]: Again, I think if we get here without gossiper we are doing something wrong.

**Comment by kbr-** [comment]
Closing, will open non-RFC version.

**Pull request closed without merging**